### ISSUE TITLE
Model overfit for any method or dataset

### ISSUE THREAD

**Issue by SanaAwan5** [JianXu95/FedPAC#5]
I am trying to run the code with both FedAvg and FedPac methods over fmnist and cifar datasets. But every time model over fits from second round. I tried changing learning rate but it's not making any difference. Where can i find the parameter settings to get optimal performance?

**Comment by JianXu95** [comment]
I think maybe setting `local_epoch`= 5 would work well, which is 0 in default in the option.py and generally should be set > 0

**Comment by SanaAwan5** [comment]
thanks for the reply. I tried that as well. Local model are training but the problem is when global model update even for once like here:

global_weight = average_weights_weighted(local_weights, agg_weight)
    # update global model
    
    global_model.load_state_dict(global_weight)

After that the model just overfit

**Comment by SanaAwan5** [comment]
I have fixed the local epochs to 5 and local iteration to 20 and getting the following output. It clearly indicates local models are training but in the second round after global update, the model stops training. Similar behavior I observed for FedPac and fmnist dataset:

<img width="567" alt="image" src="https://github.com/JianXu95/FedPAC/assets/47128808/d0f4de83-2f5e-409b-8847-6f37c5cca5d4">

**Comment by SanaAwan5** [comment]
The problem is in the following code for FedAvg. After uncommenting the line using the logic of averaging, the code works fine. Can you please confirm if that is the right way to just take weight average by dividing with number of clients instead of using agg_w at all.

def average_weights_weighted(w, avg_weight):
    """
    Returns the average of the weights.
    """
    w_avg = copy.deepcopy(w[0])
    weight = torch.tensor(avg_weight)
    agg_w = weight/(weight.sum(dim=0))
    for key in w_avg.keys():
        w_avg[key] = torch.zeros_like(w_avg[key])
        for i in range(len(w)):
            w_avg[key] += agg_w[i]*w[i][key]
        # w_avg[key] = torch.div(w_avg[key], len(w))
    return w_avg
